test/E2E/EndToEndTest.cs

```diff
@@ -9,110 +9,120 @@
 
-            TestSetup();
+            var buildCommand = new BuildCommand(TestProject, output: OutputDirectory, native: true);
 
-            TestRunCommand("dotnet", $"build --native -o {OutputDirectory}");
+            buildCommand.Execute().Should().Pass();
 
             var nativeOut = Path.Combine(OutputDirectory, "native");
-            TestOutputExecutable(nativeOut);
+            TestOutputExecutable(nativeOut, buildCommand.GetOutputExecutableName());
         }
 
         [Fact]
-        public void TestDotnetCompileNativeCpp()
+        public void TestDotnetBuildNativeCpp()
         {
-            TestSetup();
+            var buildCommand = new BuildCommand(TestProject, output: OutputDirectory, native: true, nativeCppMode: true);
+
+            buildCommand.Execute().Should().Pass();
 
-            TestRunCommand("dotnet", $"build --native --cpp -o {OutputDirectory}");
-            
             var nativeOut = Path.Combine(OutputDirectory, "native");
-            TestOutputExecutable(nativeOut);
+            TestOutputExecutable(nativeOut, buildCommand.GetOutputExecutableName());
         }
 
         [Fact]
         public void TestDotnetRun()
         {
-            TestSetup();
+            var runCommand = new RunCommand(TestProject);
 
-            TestRunCommand("dotnet", $"run");
+            runCommand.Execute()
+                .Should()
+                .Pass();
         }
         
         [Fact]
         public void TestDotnetPack()
         {
-            TestSetup();
+            var runCommand = new PackCommand(TestDirectory, output: OutputDirectory);
 
-            TestRunCommand("dotnet", $"pack");
+            runCommand.Execute()
+                .Should()
+                .Pass();
         }
 
         [Fact]
         public void TestDotnetPublish()
         {
-            TestSetup();
+            var publishCommand = new PublishCommand(TestProject, output: OutputDirectory);
+            publishCommand.Execute().Should().Pass();
 
-            TestRunCommand("dotnet", $"publish --framework dnxcore50 --runtime {Rid} -o {OutputDirectory}");
-            TestOutputExecutable(OutputDirectory);    
+            TestOutputExecutable(OutputDirectory, publishCommand.GetOutputExecutable());    
         }
 
         private void TestSetup()
         {
-            Directory.SetCurrentDirectory(RootPath);
+            var root = Temp.CreateDirectory();
+
+            TestDirectory = root.CreateDirectory(TESTDIR_NAME).Path;
+            TestProject = Path.Combine(TestDirectory, "project.json");
+            OutputDirectory = Path.Combine(TestDirectory, OUTPUTDIR_NAME);
 
-            CleanOrCreateDirectory(TestDirectory);
-            CleanOrCreateDirectory(OutputDirectory);
+            InitializeTestDirectory();   
+        }
 
+        private void InitializeTestDirectory()
+        {
+            var currentDirectory = Directory.GetCurrentDirectory();
             Directory.SetCurrentDirectory(TestDirectory);
 
```

Comment:
This is going to prevent from running tests in parallel. The `new` command must take an output directory as a parameter and that will allow us to not set the current directory.